triage_bot/lib/engine.dart

```diff
@@ -0,0 +1,1499 @@
+      thumbs: reader.readInt(),
+    );
+  }
+
+  static void write(FileWriter writer, IssueStats value) {
+    writer.writeNullOr<DateTime>(value.lastContributorTouch, writer.writeDateTime);
+    writer.writeNullOr<DateTime>(value.lastAssigneeTouch, writer.writeDateTime);
+    writer.writeSet<String>(writer.writeString, value.labels);
+    writer.writeNullOr<DateTime>(value.openedAt, writer.writeDateTime);
+    writer.writeNullOr<DateTime>(value.lockedAt, writer.writeDateTime);
+    writer.writeNullOr<DateTime>(value.assignedAt, writer.writeDateTime);
+    writer.writeBool(value.assignedToTeamMemberReporter);
+    writer.writeNullOr<DateTime>(value.triagedAt, writer.writeDateTime);
+    writer.writeNullOr<int>(value.thumbsAtTriageTime, writer.writeInt);
+    writer.writeInt(value.thumbs);
+  }
+
+  DateTime? lastContributorTouch;
+  DateTime? lastAssigneeTouch;
+  Set<String> labels;
+  DateTime? openedAt;
+  DateTime? lockedAt;
+  DateTime? assignedAt;
+  bool assignedToTeamMemberReporter = false;
+  DateTime? triagedAt;
+  int? thumbsAtTriageTime;
+  int thumbs;
+
+  @override
+  String toString() {
+    final StringBuffer buffer = StringBuffer();
+    buffer.write('{${(labels.toList()..sort()).join(', ')}} and $thumbs 👍');
+    if (openedAt != null) {
+      buffer.write('; openedAt: $openedAt');
+    }
+    if (lastContributorTouch != null) {
+      buffer.write('; lastContributorTouch: $lastContributorTouch');
+    } else {
+      buffer.write('; lastContributorTouch: never');
+    }
+    if (assignedAt != null) {
+      buffer.write('; assignedAt: $assignedAt');
+      if (assignedToTeamMemberReporter) {
+        buffer.write(' (to team-member reporter)');
+      }
+      if (lastAssigneeTouch != null) {
+        buffer.write('; lastAssigneeTouch: $lastAssigneeTouch');
+      } else {
+        buffer.write('; lastAssigneeTouch: never');
+      }
+    } else {
+      if (lastAssigneeTouch != null) {
+        buffer.write('; lastAssigneeTouch: $lastAssigneeTouch (?!)');
+      }
+      if (assignedToTeamMemberReporter) {
+        buffer.write('; assigned to team-member reporter (?!)');
+      }
+    }
+    if (lockedAt != null) {
+      buffer.write('; lockedAt: $lockedAt');
+    }
+    if (triagedAt != null) {
+      buffer.write('; triagedAt: $triagedAt');
+      if (thumbsAtTriageTime != null) {
+        buffer.write(' with $thumbsAtTriageTime 👍');
+      }
+    } else {
+      if (thumbsAtTriageTime != null) {
+        buffer.write('; not triaged with $thumbsAtTriageTime 👍 when triaged (?!)');
+      }
+    }
+    return buffer.toString();
+  }
+}
+
+typedef StoreFields = ({
+  Map<int, IssueStats> issues,
+  Map<int, DateTime> pendingCleanupIssues,
+  int? selfTestIssue,
+  DateTime? selfTestClosedDate,
+  int currentBackgroundIssue,
+  int highestKnownIssue,
+  Map<String, DateTime> lastRefeedByTime,
+  DateTime? lastCleanupStart,
+  DateTime? lastCleanupEnd,
+  DateTime? lastTidyStart,
+  DateTime? lastTidyEnd,
+});
+
+class Engine {
```

Comment:
Unit tests for this file?

Reply:
I'm open to suggestions but I really don't see much of a way to test this file. It's basically one giant ball of heuristics that either work or don't; it's not like we could refactor this code usefully and maintain the functionality. This code will certainly need continuous maintenance, but that's all about how GitHub, Discord, and our processes change over time. There's no way to write an actual spec for what this code does that has any way to be more accurate than the code.

If you have specific suggestions for how to test engine.dart please do let me know. I'd love to have more useful tests here.